File: 贪心算法/easy/122.买卖股票的最佳时机-ii.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=122 lang=cpp
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        if (prices.size() == 0)
        {
            return 0;
        }
        int sum = 0;
        int prePrice = prices[0];

        // 由于股票第二天才产生收益，所以计算收益从第二天开始
        for (int i = 1; i < prices.size(); i++)
        {
            // 当天减去前一天的价格，就是收益
            int result = prices[i] - prePrice;
            // 收益大于0就进行累加
            if (result > 0)
            {
                sum += result;
            }

            prePrice = prices[i];
        }

        return sum;
    }
};
// @lc code=end

File: 贪心算法/easy/122.买卖股票的最佳时机-ii.ts
/*
 * @lc app=leetcode.cn id=122 lang=typescript
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
function maxProfit(prices: number[]): number {
  let sum = 0

  // 假如第1天买入，第3天卖出，那么利润为：prices[2] - prices[0]
  // 相当于(prices[2] - prices[1]) + (prices[1] - prices[0])
  for (let i = 1; i < prices.length; i++) {
    // 局部最优：收集每天的正利润，把利润分解为每天为单位的维度
    if (prices[i] - prices[i - 1] > 0) {
      // 全局最优：求得最大利润
      sum += prices[i] - prices[i - 1]
    }
  }

  return sum
};
// @lc code=end

